File: CommonCode/ColumnRenderer.m
//
//  SampleSetMesh.m
//  Spectrogeddon
//
//  Created by Tom York on 22/04/2014.
//  
//

#import "ColumnRenderer.h"
#import "TimeSequence.h"
#import "ShadedMesh.h"

@interface ColumnRenderer ()
@property (nonatomic,strong) GLKTextureInfo* texture;
@property (nonatomic,strong) ShadedMesh* shadedMesh;
@property (nonatomic) BOOL invalidatedVertices;
@end


@implementation ColumnRenderer

- (void)dealloc
{
    CGImageRelease(_colorMapImage);
}

- (void)setColorMapImage:(CGImageRef)colorMapImage
{
    if(_colorMapImage != colorMapImage)
    {
        CGImageRelease(_colorMapImage);
        _colorMapImage = colorMapImage;
        CGImageRetain(colorMapImage);
        self.texture = nil;
    }
}

- (void)setUseLogFrequencyScale:(BOOL)useLogFrequencyScale
{
    if(_useLogFrequencyScale != useLogFrequencyScale)
    {
        _useLogFrequencyScale = useLogFrequencyScale;
        self.invalidatedVertices = YES;
    }
}

- (void)generateVertexPositions:(TexturedVertexAttribs* const)vertices
{
    const float logOffset = 0.001f; // Safety margin to ensure we don't try taking log2(0)
    const float logNormalization = 1.0f/log2f(logOffset);
    const float yScale = 1.0f / (float)(self.shadedMesh.numberOfVertices/2 - 1);
    for(NSUInteger valueIndex = 0; valueIndex < self.shadedMesh.numberOfVertices/2; valueIndex++)
    {
        const NSUInteger vertexIndex = valueIndex << 1;
        const float y = self.useLogFrequencyScale ? (1.0f-logNormalization*log2f((float)valueIndex*yScale+logOffset)) : (float)valueIndex * yScale;
        vertices[vertexIndex] = (TexturedVertexAttribs){ 0.0f, y, 0.0f, 0.0f };
        vertices[vertexIndex+1] = (TexturedVertexAttribs){ 1.0f, y, 0.0f, 0.0f };
    }
}

- (void)updateVerticesForTimeSequence:(TimeSequence*)timeSequence offset:(float)offset width:(float)width
{
    if(!timeSequence.numberOfValues)
    {
        return;
    }
    
    const GLsizei vertexCountForSequence = (GLsizei)(timeSequence.numberOfValues * 2);
    if(!self.shadedMesh)
    {
        self.shadedMesh = [[ShadedMesh alloc] initWithNumberOfVertices:vertexCountForSequence];
        self.invalidatedVertices = YES;
    }
    else if(self.shadedMesh.numberOfVertices != vertexCountForSequence)
    {
        [self.shadedMesh resizeMesh:vertexCountForSequence];
        self.invalidatedVertices = YES;
    }
    
    if(self.invalidatedVertices)
    {
        [self.shadedMesh updateVertices:^(TexturedVertexAttribs *const vertices) {
            [self generateVertexPositions:vertices];
        }];
        self.invalidatedVertices = NO;
    }

    [self.shadedMesh updateVertices:^(TexturedVertexAttribs *const vertices) {
        for(NSUInteger valueIndex = 0; valueIndex < timeSequence.numberOfValues; valueIndex++)
        {
            const NSUInteger vertexIndex = valueIndex << 1;
            const float value = [timeSequence valueAtIndex:valueIndex];
            vertices[vertexIndex].t = value;
            vertices[vertexIndex+1].t = value;
        }
    }];

    const GLKMatrix4 translation = GLKMatrix4MakeTranslation(offset, 0.0f, 0.0f);
    self.shadedMesh.transform = GLKMatrix4Multiply(GLKMatrix4Scale(translation, width, 1.0f, 1.0f), self.positioning);
}

- (void)render
{
    if(!self.colorMapImage || !self.shadedMesh)
    {
        return;
    }
    
    if(!self.texture && self.colorMapImage)
    {
        self.texture = [GLKTextureLoader textureWithCGImage:self.colorMapImage options:nil error:nil];
    }
    
    glBindTexture(GL_TEXTURE_2D, self.texture.name);
    [self.shadedMesh render];
    glBindTexture(GL_TEXTURE_2D, 0);
}

@end

File: CommonCode/ColumnRenderer.swift
//
//  ColumnRenderer.swift
//  SpectrogeddonOSX
//
//  Created by Tom York on 24/05/2019.
//  Copyright © 2019 Spectrogeddon. All rights reserved.
//

import GLKit

///  Renders a column, a single measurement set.
public class ColumnRenderer {
    
    public var useLogFrequencyScale: Bool {
        didSet {
            hasInvalidatedVertices = true
        }
    }

    public var colorMapImage: CGImage? {
        didSet {
            texture = nil
        }
    }

    public var positioning: GLKMatrix4
    
    private var texture: GLKTextureInfo?
    private var mesh: ShadedMesh?
    private var hasInvalidatedVertices: Bool
    
    init() {
        self.positioning = GLKMatrix4Identity
        self.hasInvalidatedVertices = false
        self.useLogFrequencyScale = false
    }
    
    func render() {
        
        guard let colorMap = colorMapImage, let mesh = mesh else {
            return
        }

        if texture == nil {
            texture = try? GLKTextureLoader.texture(with: colorMap, options: nil)
        }
        
        guard let texture = texture else {
            return
        }
        
        glBindTexture(GLenum(GL_TEXTURE_2D), texture.name)
        mesh.render()
        glBindTexture(GLenum(GL_TEXTURE_2D), 0)
    }
    
    private func loadTexture() {
        
    }
    
    func updateVertices(timeSequence: TimeSequence, offset: Float, width: Float) {
        
        guard !timeSequence.values.isEmpty else {
            return
        }
        
        let vertexCountForSequence = Int(timeSequence.values.count * 2)
        if mesh == nil {
            mesh = ShadedMesh(numberOfVertices: vertexCountForSequence)
            hasInvalidatedVertices = true
        } else if mesh?.numberOfVertices != vertexCountForSequence {
            mesh?.resize(vertexCountForSequence)
            hasInvalidatedVertices = true
        }

        if hasInvalidatedVertices {
            mesh?.updateVertices { storage in
                self.generateVertexPositions(vertices: storage)
            }
            hasInvalidatedVertices = false
        }

        mesh?.updateVertices { storage in
            
            for valueIndex in (0..<timeSequence.values.count) {
                let vertexIndex = Int(valueIndex << 1)
                let value = timeSequence.values[valueIndex]
                storage[vertexIndex].t = value
                storage[vertexIndex+1].t = value
            }
        }

        let translation = GLKMatrix4MakeTranslation(offset, 0, 0)
        mesh?.transform = GLKMatrix4Multiply(GLKMatrix4Scale(translation, width, 1, 1), positioning)
    }
    
    private func generateVertexPositions(vertices: UnsafeMutableBufferPointer<TexturedVertexAttribs>) {
        
        guard let count = mesh?.numberOfVertices else {
            return
        }
        
        let numberOfVertices = Int(count)
        
        let logOffset: Float = 0.001 // Safety margin to ensure we don't try taking log2(0)
        let logNormalization = 1/log2(logOffset)
        let yScale = 1 / Float(numberOfVertices/2 - 1)
        
        for valueIndex in (0..<numberOfVertices/2) {

            let vertexIndex = valueIndex << 1
            let y: Float
            if useLogFrequencyScale {
                y = 1.0 - logNormalization * log2(Float(valueIndex) * yScale + logOffset)
            } else {
                y = Float(valueIndex)*yScale
            }
            vertices[vertexIndex] = TexturedVertexAttribs(x: 0, y: y, s: 0, t: 0)
            vertices[vertexIndex+1] = TexturedVertexAttribs(x: 1, y: y, s: 0, t: 0)
        }
    }
}
